pkg/client/clientMutator.go

```diff
@@ -0,0 +1,82 @@
+package client
+
+import (
+	"github.com/rancher/lasso/pkg/log"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+)
+
+// Mutator is function that takes in a client and returns an new mutated client.
+type Mutator func(*Client) (*Client, error)
+
+type sharedClientFactoryWithMutation struct {
+	SharedClientFactory
+	mutator Mutator
```

Comment:
I updated the PR to remove the addition of `WithImpesonation` from controllers and factories. We don't want to support the use case for creating a SharedControllerFactory with Impersonation since impersonation requests should be more of a one-off event. If a user needs to do this, they can always create a SharedControllerFactory from a custom rest.Config.